### DISCUSSION TITLE
Unable to import mp3 in Next Js

### DISCUSSION THREAD

**Discussion by zakriabilal** [vercel/next.js#12810] [Help] [answered]
I am developing a website for that I need to import mp3 into the project but it's not letting me import it. 

I've always used Webpack's default config, not familiar with its inner workings. 

This is my next.config.js

`const withCSS = require('@zeit/next-css');
const compose = require('next-compose');
cssConfig = {};
module.exports = compose([
    [withCSS, cssConfig],
    {
        webpack: (config, options) => {
            return config
        }
    }
]);`

This is how I am importing the mp3
`
import BELL from '../../public/assets/sounds/createDrawer.mp3';
`

This is the error I am getting

`
./public/assets/sounds/createDrawer.mp3 1:0
Module parse failed: Unexpected character '�' (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
`

I am using latest Ant design, React, Redux paired with Next JS

Any help will be greatly appreciated.

**Comment by jamesmosier** [comment, marked as answer]
You could try to add a custom loader to accept your `.mp3`:

First install `file-loader` for use in webpack: `npm install file-loader --save-dev`


Then you can add the loader in your `next.config.js` to handle `.mp3` files correct.
```js
const withCSS = require('@zeit/next-css');
const compose = require('next-compose');
cssConfig = {};
module.exports = compose([
  [withCSS, cssConfig],
  {
    webpack(config, options) {
      config.module.rules.push({
        test: /\.mp3$/,
        use: {
          loader: 'file-loader',
        },
      });
      return config;
    },
  },
]);

```

or 

`npm install url-loader file-loader --save-dev`

```js
module.exports = {
  webpack(config, options) {
    const { isServer } = options;
    config.module.rules.push({
      test: /\.(ogg|mp3|wav|mpe?g)$/i,
      exclude: config.exclude,
      use: [
        {
          loader: require.resolve('url-loader'),
          options: {
            limit: config.inlineImageLimit,
            fallback: require.resolve('file-loader'),
            publicPath: `${config.assetPrefix}/_next/static/images/`,
            outputPath: `${isServer ? '../' : ''}static/images/`,
            name: '[name]-[hash].[ext]',
            esModule: config.esModule || false,
          },
        },
      ],
    });

    return config;
  },
};
```

**Reply by zakriabilal** [reply]
@jamesmosier I tried this out, it made the errors that were showing up go away but it did not get the audio working. 

**The solution that you(@jamesmosier) suggested below to @tr1s with the url-loader actually did the trick.** 

I am quoting that from your reply below.

Not sure why the file-loader is not working with mp3s in NextJS though, I would love to find that out.

> Perhaps this `next.config.js` will help:
> 
> ```js
> module.exports = {
>   webpack(config, options) {
>     const { isServer } = options;
>     config.module.rules.push({
>       test: /\.(ogg|mp3|wav|mpe?g)$/i,
>       exclude: config.exclude,
>       use: [
>         {
>           loader: require.resolve('url-loader'),
>           options: {
>             limit: config.inlineImageLimit,
>             fallback: require.resolve('file-loader'),
>             publicPath: `${config.assetPrefix}/_next/static/images/`,
>             outputPath: `${isServer ? '../' : ''}static/images/`,
>             name: '[name]-[hash].[ext]',
>             esModule: config.esModule || false,
>           },
>         },
>       ],
>     });
> 
>     return config;
>   },
> };
> ```
> 
> I was able to get it to output my mp3 from a non-public folder and linked to it via:
> 
> ```js-jsx
> import MyFile from '../assets/music.mp3';
> 
> ... 
> 
> return (
>   <>
>     <a href={MyFile} download>{MyFile}</a>
>     <audio src={MyFile} controls></audio>
>   </>
> )
> ```
> 
> That code was ported from https://github.com/twopluszero/next-images/blob/master/index.js

**Thank you for your help!**

**Reply by brokentime906** [reply]
Thank You It worked for me !!

**Reply by florian-milky** [reply]
This worked for me, especially thanks the the isServer check: `${isServer ? '../' : ''}`
Thanks!

**Reply by ghost** [reply]
Magic!!

**Reply by jotasenator** [reply]
ve to say that my code now is working very well...BUT..how is that we need to do this first in order to charge mp3 files in our next app...where can I read further? thanks for this answer

**Reply by johnbabu021** [reply]
yeah it worked

**Reply by TonyLig** [reply]
It works thanks!

**Comment by tr1s** [comment]
I was actually also having this problem, and couldn't get it to work even with a webpack config similar to @jamesmosier's above. I'll try again and see what's up, maybe I did something wrong.

**Reply by jamesmosier** [reply]
If that doesn't work, try putting the `.mp3` in the `/public` folder. For example `my-next-project/public/music.mp3`

You can then access the file like: `<audio src="/music.mp3" controls></audio>` or `<a href="/music.mp3">click</a>`

**Reply by tr1s** [reply]
Using a static file works, but that wasn't what I'm going for. I am actually building a musician website and planned to use [Howler](https://howlerjs.com/), and.. funny enough, [useSound](https://github.com/joshwcomeau/use-sound) that @leerob mentioned below is what I was trying to test when I made my initial comment on this discussion!

I tried a few things a couple days ago that didn't work, and testing some more things today and still nothing.

Using a stripped down version of the suggestion you gave:
```
module.exports = {
  webpack(config, options) {
    config.module.rules.push({
    test: /\.(ogg|mp3|wav|mpe?g)$/i,
      use: {
        loader: 'file-loader',
      },
    });
    return config;
  },
}
```

The build would work, but Next can't find the sound? and it looks like perhaps Next is using Howler under the hood?

<img width="725" alt="CleanShot 2020-05-13 at 11 31 34@2x" src="https://user-images.githubusercontent.com/19172250/81833229-75b2ff80-950d-11ea-9613-70b483529299.png">

___

This is what I was trying before, using a combination of [Nuxt's docs](https://nuxtjs.org/faq/webpack-audio-files/), [Webpack's docs](https://webpack.js.org/loaders/file-loader/), and Next's docs:

```
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    // Note: we provide webpack above so you should not `require` it
    // Perform customizations to webpack config
    // Important: return the modified config
    config.plugins.push({
      test: /\.(ogg|mp3|wav|mpe?g)$/i,
      use: [
        {
          loader: 'file-loader',
          options: {
            name: '[path][name].[ext]',
          },
        },
      ],
    })
    return config
  },
  webpackDevMiddleware: config => {
    // Perform customizations to webpack dev middleware config
    // Important: return the modified config
    return config
  },
}
```

I ended up just getting this error:

<img width="1191" alt="CleanShot 2020-05-13 at 11 37 54@2x" src="https://user-images.githubusercontent.com/19172250/81833863-3cc75a80-950e-11ea-8220-922c26522c45.png">


I just really don't know what's going on here. This seems like an opportunity for some more thorough documentation + examples. How to use and implement some common Webpack configurations could go here. What's it like to implement a multiple configs at once, etc.

Should I make an issue about this, or am I doing something wrong?

[Here's my repo](https://github.com/tr1s/nightizm-next) if you could take a look and try?

Thanks!

**Reply by jamesmosier** [reply]
Perhaps this `next.config.js` will help:

```js
module.exports = {
  webpack(config, options) {
    const { isServer } = options;
    config.module.rules.push({
      test: /\.(ogg|mp3|wav|mpe?g)$/i,
      exclude: config.exclude,
      use: [
        {
          loader: require.resolve('url-loader'),
          options: {
            limit: config.inlineImageLimit,
            fallback: require.resolve('file-loader'),
            publicPath: `${config.assetPrefix}/_next/static/images/`,
            outputPath: `${isServer ? '../' : ''}static/images/`,
            name: '[name]-[hash].[ext]',
            esModule: config.esModule || false,
          },
        },
      ],
    });

    return config;
  },
};
```

I was able to get it to output my mp3 from a non-public folder and linked to it via: 

```jsx
import MyFile from '../assets/music.mp3';

... 

return (
  <>
    <a href={MyFile} download>{MyFile}</a>
    <audio src={MyFile} controls></audio>
  </>
)
```

That code was ported from https://github.com/twopluszero/next-images/blob/master/index.js

**Reply by tr1s** [reply]
Ok so.. this worked, after I installed `url-loader`. Not exactly sure how but it worked. Confused at the stuff pointing to image folders lol. I'll keep playing around with this and trying to understand more.

Hopefully this gets easier and/or better documented in the future.

Thanks!

**Reply by jamesmosier** [reply]
You could probably change it to be something other than images, but I haven’t played with that so I’m unsure. Glad it worked though!

**Reply by tr1s** [reply]
So, I stripped everything back as far as I could.. turns out `file-loader` just doesn't work at all in this case. `url-loader` does the trip.

My final output:
```
module.exports = {
  webpack(config, options) {
    config.module.rules.push({
      test: /\.(ogg|mp3|wav|mpe?g)$/i,
      use: [
        {
          loader: 'url-loader',
          options: {
            name: '[name]-[hash].[ext]',
          },
        },
      ],
    });
    return config;
  },
};
```

Much cleaner :)

I wonder why file-loader didn't work.

**Reply by chetan-vc** [reply]
@jamesmosier Thanks, this helped me.

**Reply by msaaddev** [reply]
@jamesmosier thanks! You are a life saver.

**Reply by MdWahiduzzamanEmon** [reply]
It works fine. Thanks, bro

**Comment by leerob** [comment]
This might be relevant: `useSound` [React hook](https://joshwcomeau.com/react/announcing-use-sound-react-hook/).

**Reply by tr1s** [reply]
I actually was trying this before you mentioned, with no luck. Check out my [reply above](https://github.com/zeit/next.js/discussions/12810#discussioncomment-13840)!

Would love to know if you've gotten it to work?

**Comment by gajus** [comment]
In latest Next.js you need to provide additional public path configuration, e.g.

```js
config.module.rules.push({
  test: /\.(mp3)$/,
  use: {
    loader: 'file-loader',
    options: {
      publicPath: '/_next/static/sounds/',
      outputPath: 'static/sounds/',
      name: '[name].[ext]',
      esModule: false,
    },
  },
});

```

**Comment by hnapoles** [comment]
This is working for me on nextjs 10.0.5. 
/const withCSS = require('@zeit/next-css');
const compose = require('next-compose');
//cssConfig = {};
module.exports = compose([
  //[withCSS, cssConfig],
  {
    webpack(config, options) {
      config.module.rules.push({
        test: /\.mp3$/,
        use: {
          loader: 'file-loader',
        },
      });
      return config;
    },
  },
]);

import useSound from 'use-sound';
...
const [playSound] = useSound(
    './sounds/pop3.mp3',
    { volume: 0.25 }
  );
 
  const handleSubmit = (e) => {
    e.preventDefault()
    playSound()
  }

**Comment by iqraabdulrauf** [comment]
Thanks all this worked like a charm :)

**Comment by binajmen** [comment]
In my case, I had to additionally update `next-env.d.ts` with
```js
declare module '*.mp3' {
  const src: string;
  export default src;
}
```
In order to use
```js
import notification from "../../public/static/notification.mp3"
```

**Reply by timini** [reply]
Tried this but it seems you shouldnt edit the `next-env.d.ts` file directly as it is autogenerated by next.js, your changes will be lost.

```
Instead of editing next-env.d.ts, you can include additional types by adding a new file e.g. additional.d.ts and then referencing it in the include array in your tsconfig.json.
```

https://nextjs.org/docs/basic-features/typescript

**Reply by kirso** [reply]
Same here @binajmen, tried to update but the changes were adjusted automatically.

**Comment by panathea** [comment]
With Webpack 5 support, the solution becomes even more terse:

```js
// next.config.js
module.exports = {
  webpack(config, options) {
    config.module.rules.push({
      test: /\.(mp3)$/,
      type: "asset/resource",
      generator: {
        filename: "static/chunks/[path][name].[hash][ext]",
      },
    });

    return config;
  },
};
```

If you are using TypeScript, add a **new** definition file (I put mine next to `next-env.d.ts`, and the name doesn't matter:

```js
// custom-env.d.ts
declare module '*.mp3' {
  const src: string;
  export default src;
}
```

**Reply by Pipe-Runner** [reply]
Thanks a lot. This worked nicely for me. And it works both for public and non-public directory

**Reply by timini** [reply]
This worked for local dev but i get an error when trying to build

**Reply by timini** [reply]
It was typescript error. Use binajmens fix.

**Reply by yume-chan** [reply]
For me it still doesn't work, unless I remove `[path]` from `filename`. Maybe because I'm importing a symlink and there is some bug.

**Reply by panathea** [reply]
Updated to add that the file should *not* be next-env.d.ts, but rather any other definition file so that it doesn't get overwritten when you run install or build the application.

**Reply by Alex23rodriguez** [reply]
I think this is the best answer :) it is more concise and it doesn't depend on third party packages! (namely `file-loader` and `url-loader`), plus it also works for `next.config.mjs` files, whereas the marked answer requires one to change to `.js` because of the `require.resolve`, or at least that was my experience using `vite` / [create-t3-app](https://github.com/t3-oss/create-t3-app) and [UIfx](https://github.com/wle8300/uifx) to load in the audio

**Reply by sahandghorbani** [reply]
/ NOTE: This file should not be edited

**Comment by iamzapata** [comment]
If you're trying this [useSound hook](https://github.com/joshwcomeau/use-sound) with NextJS, all you really need to do is (as of April 2022): 

1) add the type declaration that @binajmen mentioned:
```javascript
declare module '*.mp3' {
  const content: string
  export default content
}

```
2) In the hook, make sure you **inline** the path to the sound file instead of using an **import** statement: 

```javascript
const [playWow] = useSound('/sounds/wow.mp3')
```

**Reply by rafaelvf** [reply]
I don't understand how to add this declaration. This is what I have  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],

**Reply by iamzapata** [reply]
@rafaelvf so you need a new definition file if you don't already have one. 

<img width="315" alt="Screenshot 2022-04-22 at 05 31 10" src="https://user-images.githubusercontent.com/60562158/164590867-b362940a-229c-433d-8eda-a2d289798fe4.png">
<img width="790" alt="Screenshot 2022-04-22 at 05 31 45" src="https://user-images.githubusercontent.com/60562158/164590868-fd86f625-2ccf-40c7-acfb-a540acb6ce8f.png">

**Reply by nos64** [reply]
Thanks a lot!

**Reply by mattp0123** [reply]
What's the purpose of the sounds.d.ts file if you are not going to use import statement?

**Reply by adarshrao** [reply]
Thanks the 2nd point was very helpful. 

`const [playWow] = useSound('/sounds/wow.mp3')`

**Comment by nikelborm** [comment]
Is there a way to do this with turbopack?

**Reply by thiisiriya** [reply]
With turbopack you can setup like this

```
  turbopack: {
    rules: {
      '*.mp3': {
        loaders: [
          {
            loader: 'url-loader',
            options: {
              name: '[name]-[hash].[ext]'
            }
          }
        ],
        as: '*.js'
      }
    }
  }
```
